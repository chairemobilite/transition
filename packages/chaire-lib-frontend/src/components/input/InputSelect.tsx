/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React, { JSX } from 'react';

export type choiceType = {
    value: string;
    disabled?: boolean;
    label?: string;
    choices?: choiceType[];
    [key: string]: unknown;
};

export type InputSelectProps = {
    id: string;
    onValueChange?: (e: any) => void;
    value?: string;
    defaultValue?: string;
    noBlank?: boolean;
    choices?: choiceType[];
    localePrefix?: string;
    t?: (string) => string;
    disabled?: boolean;
};

type defaultInputType = {
    name: string;
    id: string;
    value?: string;
    defaultValue?: string;
    onChange?: (e: any) => void;
    disabled?: boolean;
};

const InputSelect: React.FunctionComponent<InputSelectProps> = ({
    id,
    onValueChange = undefined,
    value = undefined,
    defaultValue = '',
    noBlank = false,
    choices = [],
    localePrefix = 'main',
    t = (_str) => '',
    disabled = false
}: InputSelectProps) => {
    const actualValue = value === undefined ? defaultValue : value;
    const defaultAttributes: defaultInputType = {
        name: id,
        id: id,
        value: actualValue
    };

    if (onValueChange) {
        defaultAttributes.onChange = onValueChange;
    }

    if (disabled === true) {
        defaultAttributes.disabled = true;
    }

    const selectChoices: JSX.Element[] = [];
    for (let i = 0, count = choices.length; i < count; i++) {
        const choice = choices[i];
        if (choice.choices) {
            const childChoices = choice.choices.map((childChoice) => {
                return (
                    <option key={`${childChoice.value}`} value={childChoice.value} disabled={choice.disabled}>
                        {childChoice.label || t(`${localePrefix}:${childChoice.value}`)}
                    </option>
                );
            });
            selectChoices.push(
                <optgroup key={`group_${choice.value}`} label={choice.label || t(`${localePrefix}:${choice.value}`)}>
                    {childChoices}
                </optgroup>
            );
        } else {
            selectChoices.push(
                <option key={`${choice.value}`} value={choice.value} disabled={choice.disabled}>
                    {choice.label || t(`${localePrefix}:${choice.value}`)}
                </option>
            );
        }
    }

    return (
        <select
            autoComplete="none"
            className="tr__form-input-select apptr__input _input-select"
            {...defaultAttributes}
            style={{
                overflow: 'hidden',
                whiteSpace: 'nowrap',
                textOverflow: 'ellipsis'
            }}
        >
            {!noBlank && <option key={'_blank'} value=""></option>}
            {selectChoices}
        </select>
    );
};

export default InputSelect;
