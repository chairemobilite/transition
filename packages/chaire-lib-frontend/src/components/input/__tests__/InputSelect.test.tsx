/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import InputSelect from '../InputSelect';

const mockOnChange = jest.fn();
const testId = 'SelectWidgetId';
const testLabel = 'Select label';
const defaultChoiceValue = 'test1';
const anotherChoiceValue = 'test2';
const testChoices = [
    { value: defaultChoiceValue },
    { value: anotherChoiceValue, label: 'label for test2' },
    { value: 'disabled', disabled: true },
    { value: 'has sub-choices',
        choices: [
            { value: 'subChoice1' },
            { value: 'subChoice2', label: 'label for subchoice2' }
        ]
    }
];

test('Default props', () => {
    const { container } = render(<InputSelect
        id = {testId}
    />);
    expect(container).toMatchSnapshot();
});

test('All props', () => {
    const { container } = render(<InputSelect
        id = {testId}
        onValueChange = {mockOnChange}
        value = {anotherChoiceValue}
        defaultValue = {defaultChoiceValue}
        disabled = {false}
        noBlank = {true}
        choices = {testChoices}
        localePrefix = "something"
        t = {(str) => str}
    />);
    expect(container).toMatchSnapshot();
});

test('Disabled', () => {
    const { container } = render(<InputSelect
        id = {testId}
        onValueChange = {mockOnChange}
        disabled = {true}
    />);
    expect(container).toMatchSnapshot();
});

test('With blank choice', () => {
    const { container } = render(<InputSelect
        id = {testId}
        onValueChange = {mockOnChange}
        noBlank = {false}
        choices = {testChoices}
    />);
    expect(container).toMatchSnapshot();
});

test('Default value', () => {
    mockOnChange.mockClear();
    const { getByLabelText } = render(
        <div>
            <InputSelect
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = {defaultChoiceValue}
                choices = {testChoices}
            />
            <label htmlFor={testId}>{testLabel}</label>
        </div>);
    const input = getByLabelText(testLabel) as HTMLInputElement;
    expect(input.value).toBe(defaultChoiceValue);
});

test('Default and initial value', () => {
    mockOnChange.mockClear();
    const { getByLabelText } = render(
        <div>
            <InputSelect
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = {defaultChoiceValue}
                value = {anotherChoiceValue}
                choices = {testChoices}
            />
            <label htmlFor={testId}>{testLabel}</label>
        </div>);
    const input = getByLabelText(testLabel) as HTMLInputElement;
    expect(input.value).toBe(anotherChoiceValue);
});

test('Default and initial empty value', () => {
    mockOnChange.mockClear();
    const value = '';
    const { getByLabelText } = render(
        <div>
            <InputSelect
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = {defaultChoiceValue}
                value = {value}
                choices = {testChoices}
            />
            <label htmlFor={testId}>{testLabel}</label>
        </div>);
    const input = getByLabelText(testLabel) as HTMLInputElement;
    expect(input.value).toBe(value);
});

test('Invalid default and value', () => {
    mockOnChange.mockClear();
    const { getByLabelText } = render(
        <div>
            <InputSelect
                id = {testId}
                onValueChange = {mockOnChange}
                defaultValue = "not a value"
                value = "still not a choice"
                choices = {testChoices}
                noBlank = {true}
            />
            <label htmlFor={testId}>{testLabel}</label>
        </div>);
    const input = getByLabelText(testLabel) as HTMLInputElement;
    expect(input.value).toBe(defaultChoiceValue);
});

test('Call onChange', () => {
    mockOnChange.mockClear();
    const { getByLabelText } = render(
        <div>
            <InputSelect
                id = {testId}
                onValueChange = {mockOnChange}
                choices = {testChoices}
            />
            <label htmlFor={testId}>{testLabel}</label>
        </div>);
    const input = getByLabelText(testLabel) as HTMLInputElement;
    const newText = 'new text';
    fireEvent.change(input, { target: { value: newText } });
    expect(mockOnChange).toHaveBeenCalledTimes(1);
});
