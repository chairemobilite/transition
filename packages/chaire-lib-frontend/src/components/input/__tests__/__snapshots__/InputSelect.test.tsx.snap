// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All props 1`] = `
<select
  autoComplete="none"
  className="tr__form-input-select apptr__input _input-select"
  id="SelectWidgetId"
  name="SelectWidgetId"
  onChange={[MockFunction]}
  style={
    Object {
      "overflow": "hidden",
      "textOverflow": "ellipsis",
      "whiteSpace": "nowrap",
    }
  }
  value="test2"
>
  <option
    value="test1"
  >
    something:test1
  </option>
  <option
    value="test2"
  >
    label for test2
  </option>
  <option
    disabled={true}
    value="disabled"
  >
    something:disabled
  </option>
  <optgroup
    label="something:has sub-choices"
  >
    <option
      value="subChoice1"
    >
      something:subChoice1
    </option>
    <option
      value="subChoice2"
    >
      label for subchoice2
    </option>
  </optgroup>
</select>
`;

exports[`Default props 1`] = `
<select
  autoComplete="none"
  className="tr__form-input-select apptr__input _input-select"
  id="SelectWidgetId"
  name="SelectWidgetId"
  style={
    Object {
      "overflow": "hidden",
      "textOverflow": "ellipsis",
      "whiteSpace": "nowrap",
    }
  }
  value=""
>
  <option
    value=""
  />
</select>
`;

exports[`Disabled 1`] = `
<select
  autoComplete="none"
  className="tr__form-input-select apptr__input _input-select"
  disabled={true}
  id="SelectWidgetId"
  name="SelectWidgetId"
  onChange={[MockFunction]}
  style={
    Object {
      "overflow": "hidden",
      "textOverflow": "ellipsis",
      "whiteSpace": "nowrap",
    }
  }
  value=""
>
  <option
    value=""
  />
</select>
`;

exports[`With blank choice 1`] = `
<select
  autoComplete="none"
  className="tr__form-input-select apptr__input _input-select"
  id="SelectWidgetId"
  name="SelectWidgetId"
  onChange={[MockFunction]}
  style={
    Object {
      "overflow": "hidden",
      "textOverflow": "ellipsis",
      "whiteSpace": "nowrap",
    }
  }
  value=""
>
  <option
    value=""
  />
  <option
    value="test1"
  >
    
  </option>
  <option
    value="test2"
  >
    label for test2
  </option>
  <option
    disabled={true}
    value="disabled"
  >
    
  </option>
  <optgroup
    label=""
  >
    <option
      value="subChoice1"
    >
      
    </option>
    <option
      value="subChoice2"
    >
      label for subchoice2
    </option>
  </optgroup>
</select>
`;
